eurec4a_environment/variables/boundary_layer/mixed_layer_height.py

```diff
@@ -93,3 +91,79 @@ def calc_peakRH_linearize(
     da.attrs["long_name"] = "mixed layer height (from RH peak and linearization)"
     da.attrs["units"] = "m"
     return da
+
+
+def calc_from_gradient(
+    ds, var, threshold, z_min=200, altitude="height", time="sounding"
+):
+    """
+    Find mixed layer height as layer over which x(z+1) - x(z) < threshold
+
+    Inputs:
+        -- ds: Dataset
+        -- var: variable name
+        -- threshold: in units of variable (i.e. g/kg or K). Maximum difference allowed in one vertical step, dz=10m, i.e. threshold_q = 0.4 g/kg,
+            or vertical gradient dq/dz < 0.04 g/kg/m
+
+    Outputs: DataArray containing mixed layer height from gradient method
+
+    Note that function is slow and should be optimized
+    """
+
+    def calculateHmix_var(
+        density_profile,
+        var_profile,
+        threshold,
+        z_min,
+        altitude="height",
+        time="sounding",
+    ):
+
+        var_diff = 0
+        numer = 0
+        denom = 0
+        var_mix = 0
+
+        k = int(z_min / 10)  # enforce lower limit = 200m. k is index of lower limit
+
+        # var(z) - weighted_mean(z-1) < threshold
+        while abs(var_diff) < threshold:
+
+            numer += 0.5 * (
+                density_profile[k + 1] * var_profile[k + 1]
+                + density_profile[k] * var_profile[k]
+            )
+            denom += 0.5 * (density_profile[k + 1] + density_profile[k])
+            var_mix = numer / denom
+            k += 1
+            var_diff = (var_profile[k] - var_mix).values
+
+        hmix = var_profile[altitude].values[k]
+        return hmix
+
+    # call inner function
+    from eurec4a_environment.variables.calc_density import calc_density
+
+    da_density = calc_density(ds)
+
+    hmix_vec = np.zeros(len(ds[var]))
```

Comment:
This looks good! I'm wondering how to do something similar to the for-loop to load individual variable and density profiles?
`da_height = ds.stack(dict(n=dims)).groupby("n").apply(_calc_height).unstack("n")`
vs.
`hmix_vec = np.zeros(len(ds[var]))
    for i in range((len(ds[var]))):
        density_profile = da_density.isel({time: i})
        var_profile = ds[var].isel({time: i})
        hmix_vec[i] = _calc_height(
            density_profile,
            var_profile,
            threshold,
            z_min,
            altitude="height",
            time="sounding",
        )`

Reply:
Yes, I think this looks good. I hadn't really considered this earlier, but most of the calculations we do operate on a single-column at a time, so this is probably a general pattern we want to follow. Where the calculation only works if a single column is used at a time we can use `stack(..).groupby(...).apply(...).unstack(...)`.

Also, where you're defining a nested function (here `_calc_height`) inside another function you don't need to pass in the variables already available (for example `z_min`). Actually it's best to avoid doing that in case you accidentally redefine what a variable means within a function (by passing in the wrong thing for for example `z_min` or setting it to something else in the call).